src/openqaoa-core/algorithms/qaoa/qaoa_workflow.py

```diff
@@ -301,11 +321,107 @@ def optimize(self, verbose=False):
+
+        # before evaluating the circuit we check that the QAOA object has been compiled
+        if self.compiled is False:
+            raise ValueError("Please compile the QAOA before optimizing it!")
+
+        # Check the type of the input parameters and save them as a
+        # QAOAVariationalBaseParams object at the variable `params_obj`
+
+        # if the parameters are passed as a dictionary we copy and update the variational parameters of the QAOA object
+        if isinstance(params, dict):
+            params_obj = deepcopy(self.variate_params)
+            # we check that the dictionary contains all the parameters of the QAOA object that are not empty
+            for key, value in params_obj.asdict().items():
+                if value.size > 0:
+                    assert (
+                        key in params.keys()
+                    ), f"The parameter `{key}` is missing from the input dictionary"
+            params_obj.update_from_dict(params)
+
+        # if the parameters are passed as a list we copy and update the variational parameters of the QAOA object
+        elif isinstance(params, list) or isinstance(params, np.ndarray):
+            assert len(params) == len(
+                self.variate_params
+            ), "The number of parameters does not match the number of parameters in the QAOA circuit"
+            params_obj = deepcopy(self.variate_params)
+            params_obj.update_from_raw(params)
+
+        # if the parameters are passed as a QAOAVariationalBaseParams object we just take it as it is
+        elif isinstance(params, QAOAVariationalBaseParams):
+            # check whether the input params object is supported for circuit evaluation
+            assert (
+                len(self.variate_params.mixer_1q_angles) == len(params.mixer_1q_angles)
+                and len(self.variate_params.mixer_2q_angles)
+                == len(self.variate_params.mixer_2q_angles)
+                and len(self.variate_params.cost_1q_angles)
+                == len(self.variate_params.cost_1q_angles)
+                and len(self.variate_params.cost_2q_angles)
+                == len(self.variate_params.cost_2q_angles)
+            ), "Specify a supported params object"
+            params_obj = params
+
+        # if the parameters are passed in a different format, we raise an error
+        else:
+            raise TypeError(
+                f"The input params must be a list or a dictionary. Instead, received {type(params)}"
+            )
+
+        # Evaluate the QAOA circuit and return the results
+        output_dict = {
+            "cost": None,
+            "uncertainty": None,
+            "measurement_results": None,
+        }
+        # if the backend is the analytical simulator, we just return the expectation value of the cost Hamiltonian
+        if isinstance(self.backend, QAOABackendAnalyticalSimulator):
+            output_dict.update({"cost": self.backend.expectation(params_obj)[0]})
```

Comment:
I have the feeling we are not routing the circuit in this function. @vishal-ph could you confirm if everything looks good here?